src/ripple/overlay/impl/PeerReservationTable.cpp

```diff
@@ -0,0 +1,168 @@
+PeerReservation::toJson() const -> Json::Value
+{
+    Json::Value result{Json::objectValue};
+    result[jss::node] = toBase58(TokenType::NodePublic, nodeId);
+    if (!description.empty())
+    {
+        result[jss::description] = description;
+    }
+    return result;
+}
+
+auto
+PeerReservationTable::list() const -> std::vector<PeerReservation>
+{
+    std::vector<PeerReservation> list;
+    list.reserve(table_.size());
+    {
+        std::lock_guard<std::mutex> lock(mutex_);
+        std::copy(table_.begin(), table_.end(), std::back_inserter(list));
+    }
+    std::sort(list.begin(), list.end());
+    return list;
+}
+
+// See `ripple/app/main/DBInit.cpp` for the `CREATE TABLE` statement.
+// It is unfortunate that we do not get to define a function for it.
+
+// We choose a `bool` return type to fit in with the error handling scheme
+// of other functions called from `ApplicationImp::setup`, but we always
+// return "no error" (`true`) because we can always return an empty table.
+bool
+PeerReservationTable::load(DatabaseCon& connection)
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    connection_ = &connection;
+    auto db = connection_->checkoutDb();
+
+    boost::optional<std::string> valPubKey, valDesc;
+    // We should really abstract the table and column names into constants,
+    // but no one else does. Because it is too tedious? It would be easy if we
+    // had a jOOQ for C++.
+    soci::statement st =
+        (db->prepare << "SELECT PublicKey, Description FROM PeerReservations;",
+         soci::into(valPubKey),
+         soci::into(valDesc));
+    st.execute();
+    while (st.fetch())
+    {
+        if (!valPubKey || !valDesc)
+        {
+            // This represents a `NULL` in a `NOT NULL` column. It should be
+            // unreachable.
+            continue;
+        }
+        auto const optNodeId =
+            parseBase58<PublicKey>(TokenType::NodePublic, *valPubKey);
+        if (!optNodeId)
+        {
+            JLOG(journal_.warn()) << "load: not a public key: " << valPubKey;
+            continue;
+        }
+        table_.insert(PeerReservation{*optNodeId, *valDesc});
+    }
+
+    return true;
+}
+
+auto
+PeerReservationTable::insert_or_assign(
+        PeerReservation const& reservation)
+    -> boost::optional<PeerReservation>
+{
+    boost::optional<PeerReservation> previous;
+
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    auto hint = table_.find(reservation);
+    if (hint != table_.end()) {
+        // The node already has a reservation. Remove it.
+        // `std::unordered_set` does not have an `insert_or_assign` method,
+        // and sadly makes it impossible for us to implement one efficiently:
+        // https://stackoverflow.com/q/49651835/618906
+        // Regardless, we don't expect this function to be called often, or
+        // for the table to be very large, so this less-than-ideal
+        // remove-then-insert is acceptable in order to present a better API.
```

Comment:
FWIW, I suspect the awkwardness of this method suggests that the right choice for `table_`'s container would have been a `map` or `unordered_map`.  Then you'd be able to replace the `value` without needing to remove the entire node.  But I also know this has been discussed plenty, and I'm not asking for any changes.

Reply:
I disagree. I think it's a problem with `unordered_set`. If it had an `insert_or_assign` method the same as `unordered_map`, then there would be no awkwardness. There is no technical reason it cannot have such a method. I don't think a deficiency in one standard container means the right choice is to pick a differently deficient container.

Reply:
From the perspective of the standard designers there is no deficiency.  The `key` to both a `set` and a `map` _must_ be immutable.  If you were allowed to mutate a `key` in place that would corrupt the data structure.  So, sensibly enough, the interface doesn't allow you to mutate the `key` (without working hard).

A `set` _only_ has a `key`, and there are no options for a partially non-`const` `key`.  If you need a partially non-`const` key, well, that's what `map` is for.

Reply:
Can you define "mutate the key"? I don't see it as a mutation of the key. I see it as an atomic replacement. It is not like asking the container to return an iterator to a non-const key where I can mutate the key outside the purview of the container. I am asking the container itself to atomically remove one key and insert another. It has full control of its invariants before, during, and after that process. No corruption would be possible. I don't need the `key_type` or `value_type` of `unordered_set` to be mutable at all. I just need an amotic `replace`/`upsert`/`insert_or_assign`/whatever-you-want-to-call-it method.

Reply:
The reason I was referring to "mutate the key" is because the part of the key that determines the location in the `set` does not change during the "atomic replacement".  The only thing that changes is the `description_`.  So you could leave the key in place and "mutate the key" by changing (only) `description_`.

I'm not going to argue with you about whether `insert_or_assign` _should_ be a part of the standard `set` interface.